doc/developer/nxeps/nxep-0002.rst

```diff
@@ -0,0 +1,109 @@
+==================================
+NXEP 2 — API design of view slices
+==================================
+
+:Author: Mridul Seth
+:Status: Draft
+:Type: Standards Track
+:Created: 2020-07-23
+
+
+Abstract
+--------
+
+The core API of base classes in NetworkX uses (dict) views to expose
+data about the graph objects. In this NXEP we propose to add slicing to Node(Data)View
+and Edge(Data)View.
+
+Motivation and Scope
+--------------------
+
+While accessing Graph data with G.nodes and G.edges, the only way of slicing the data
+is by casting the view to a list manually and then calling a slice on it.
+
+``G.nodes(data=True)`` returns a NodeDataView of all the nodes, ``G.nodes(data=True)[x]`` returns an attribute dictionary for the node x.
+The current way of getting a slice out of the underlying dict view is to cast it to list and then
+slice it ``list(G.nodes(data=True))[0:10]``. This bit of code is something that is written a lot of times
+by users. For graphs with a lot of nodes and edges, ``G.nodes`` and ``G.edges`` will take a lot of screen space and
+when the users try to slice the resulting view (the first instinct) it will error out. Users definitely need to go through
+a couple of documentation links before they realise that they need to first cast this NodeDataView to a list and then create
+a slice. In this NXEP we propose to move the ``list(G.nodes(data=True))[0:10]`` part of code inside ``__getitem__`` so that users
+can get the small subset of nodes by creating a slice ``G.nodes(data=True)[0:10]`` or a new slicing method like ``G.nodes(data=True).slice(0, 10)``
+to make the API less ambigious.
+
+Usage and Impact
+----------------
+
+The main impact and the decision that needs to be taken in this NXEP is with
+respect to the user facing API. By implemeting this NXEP we may end up adding some
+ambiguity for users. As for example `G.nodes[x]` will return an attribute dict
+but `G.nodes[0:5]` will return a list of first five nodes. This will be more ambigious with
+EdgeView as ``G.edges[0, 1]`` will return an attribute dictionary of the edge between 0 and
+1 and ``G.edges[0:1]`` will return the first edge.
+We need to find a way to counter this.
+
+For a historical context, in pre 2.0 NetworkX slicing was native behavior as G.nodes() and G.edges()
+were lists and list of tuples respectively. With this NXEP we would like to bring the intuitiveness
+of slicing behavior back to G.edges and G.nodes.
```

Comment:
> Was there any conversation at the time on how this would affect user-facing behavior?

I am not sure if there was a discussion *specifically* regarding this, @dschult would know better.

Reply:
I added some historical context to the document. We did not expect people to slice the lists... In the old system, the lists were not consistent from call to call if the graph had changed and the dicts reshuffled the order of the nodes would be different.  If you wanted only the first 10 you would use the iterator and stop after 10.

The intent of the changes was 
1) to reduce the number of methods to access the nodes from 3 to 1 
2) to construct a Python 3 dict-like interface with views (and their iterators) rather than lists.

We didn't even consider G.nodes to be an ordered container (sequence) that could be sliced. That sort of happened when Python 3.6 made the dicts ordered.